subprojects/android-test/toast-rule/src/main/kotlin/com/avito/android/rule/internal/ToastTextAccessor.kt

```diff
@@ -0,0 +1,59 @@
+package com.avito.android.rule.internal
+
+import android.os.Build
+import android.widget.TextView
+import android.widget.Toast
+import androidx.test.platform.app.InstrumentationRegistry
+import com.avito.android.test.util.HiddenApiOpener
+
+internal interface ToastTextAccessor {
+
+    /**
+     * Reads text from Toast created by [Toast.makeText].
+     * Returns null in case of custom view
+     */
+    fun text(toast: Toast): String?
+
+    companion object {
+
+        fun create(): ToastTextAccessor {
+            val inAppToasts = Build.VERSION.SDK_INT < 30 ||
+                InstrumentationRegistry.getInstrumentation().targetContext.applicationInfo.targetSdkVersion < 30
+
+            return if (inAppToasts) {
+                InAppToastTextAccessor()
+            } else {
+                SystemUIToastTextAccessor()
+            }
+        }
+    }
+}
+
+internal class InAppToastTextAccessor : ToastTextAccessor {
+
+    override fun text(toast: Toast): String? {
+        @Suppress("DEPRECATION")
+        val view = requireNotNull(toast.view) {
+            "Toast.view must be present in $toast"
+        }
+        val messageView: TextView? = view.findViewById(android.R.id.message)
+        val isCustomView = messageView == null
+        return if (isCustomView) null else messageView!!.text.toString()
+    }
+}
+
+internal class SystemUIToastTextAccessor : ToastTextAccessor {
+
+    override fun text(toast: Toast): String? {
```

Comment:
Let's discuss a behavior.  We are trying to express the complex state via nullabillity

Reply:
Added an explicit check for the existence of the field.
Besides that, it looks ok. The field is nullable in the case of a custom view.